### PR TITLE
Add support for 3rd party providers to `gr.load`, and provide a better UX for conversational models

### PR DESCRIPTION
**Pull request by abidlabs**
After discussions with @AK391 how to optimize `gr.load` for potential integrations with 3rd party model providers, I've made some (breaking) changes:

1. Renamed `hf_token` to `token` (while soft-deprecating the former)
2. Modified `src` to be able to take any user-defined function that accepts a model name and token and returns a Gradio app. I've modified the docstring of `gr.load` accordingly. The idea is that we want to be able support e.g.

```
import gradio as gr
import lepton_gradio

gr.load("models/meta-llama/Llama-3.2-11B-Vision-Instruct", src=lepton_gradio.registry).launch()
```

(as @pngwn suggested, we use the registry approach instead of monkey-patching `gr.load`)

3. Partially reverted https://github.com/gradio-app/gradio/pull/9069, in that we _do_ pass in a locally-saved `hf_token` to `gr.load` in the case of model repos, but not in any other case.
4. Finally, when a conversational model is loaded with `gr.load`, we now produce a `gr.ChatInterface`, not a `gr.Interface` as this is a much better UX. Example:

```
import gradio as gr

gr.load("meta-llama/Meta-Llama-3-8B-Instruct", src="models").launch()
```

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@ fastapi<1.0
 ffmpy
 gradio_client==1.4.0-beta.2
 httpx>=0.24.1
-huggingface_hub>=0.19.3
+huggingface_hub>=0.22.0
```
Comment (abidlabs) [inline requirements.txt:7]: When `.chat_completion` was added

File: gradio/external.py [hunk 2]
```diff
@@ -37,93 +36,91 @@
...
+        src: function that accepts a string model `name` and a string or None `token` and returns a Gradio app. Alternatively, this parameter takes one of two strings for convenience: "models" (for loading a Hugging Face model through the Inference API) or "spaces" (for loading a Hugging Face Space). If None, uses the prefix of the `name` parameter to determine `src`.
+        token: optional token that is passed as the second parameter to the `src` function. For Hugging Face repos, uses the local HF token when loading models but not Spaces (when loading Spaces, only provide a token if you are loading a trusted private Space as the token can be read by the Space you are loading). Find HF tokens here: https://huggingface.co/settings/tokens.
+        kwargs: additional keyword parameters to pass into the `src` function. If `src` is "models" or "Spaces", these parameters are passed into the `gr.Interface` or `gr.ChatInterface` constructor.
```
Comment (freddyaboulton) [inline gradio/external.py:52]: I would rather let the `src` function accept kwargs. the current `model_name, token` signature is too strict

**Review by freddyaboulton — APPROVED** [review]
Looks good @abidlabs ! Just a small suggestion on the callable's signature

File: gradio/external.py (hunk 2 above)
Comment (abidlabs) [reply gradio/external.py:52]: Sorry wdym exactly? The Callable also accepts kwargs along with name and token.

Comment (abidlabs) [reply gradio/external.py:52]: The name and token are always passed in as they are just forwarded from gr.load

Comment (abidlabs) [reply gradio/external.py:52]: Discussed with @freddyaboulton and we are good to go here

**Pull request merged**